overlore/llm/open_ai.py

```diff
@@ -0,0 +1,172 @@
+        passive_realm_entity_id = int(str(event[3]))
+        active_realm_name = realms.name_by_id(active_realm_entity_id)
+        passive_realm_name = realms.name_by_id(passive_realm_entity_id)
+
+        metadata: dict[Any, Any] = str_to_json(str(event[6]))
+        if metadata["damage"] == 0:
+            nl += f"Pillage of realm {passive_realm_name} by realm {active_realm_name}. "
+            stolen_resources = self._resources_to_nl(metadata["stolen_resources"])
+            nl += f"Stolen resources are {stolen_resources}."
+        else:
+            nl += f"War waged: by {active_realm_name} against {passive_realm_name}. "
+            winner = metadata["winner"]
+            winner = active_realm_name if winner == Winner.Attacker.value else passive_realm_name
+            loser = passive_realm_name if winner == Winner.Attacker.value else active_realm_name
+            nl += f"Winner is {winner}. Loser is {loser}. "
+            nl += f"Damages taken by loser: {metadata['damage']}. "
+        nl += "\n"
+        return nl
+
+    def npcs_to_nl(self, villagers: list[Villager]) -> str:
+        return "\n".join(self._npc_to_nl(npc) for npc in villagers)
+
+    def _events_to_nl(self, events: list[StoredEvent]) -> str:
+        return "\n".join(self._event_to_nl(event) for event in events)
+
+    def _event_to_nl(self, event: StoredEvent) -> str:
+        #  event
+        #  0      1     2                       3                        4           5   6         70         1          80          1
+        # [rowid, type, active_realm_entity_id, passive_realm_entity_id, importance, ts, metadata, (X_active, Y_active), (X_passive, Y_passive)]
+        event_type = event[1]
+        if event_type == EventType.COMBAT_OUTCOME.value:
+            return self._combat_outcome_to_nl(event=event)
+        elif event_type == EventType.ORDER_ACCEPTED.value:
+            return self._order_accepted_to_nl(event=event)
+        else:
+            raise RuntimeError("Unknown event type")
+
+    async def request_prompt(self, system: str, user: str) -> str:
+        response = self.client.chat.completions.create(
+            model="gpt-4-1106-preview",
+            messages=[
+                {"role": "system", "content": system},
+                {"role": "user", "content": user},
+            ],
+        )
+        if response is None:
+            raise RuntimeError("GPT interface could not request a townhall")
+        return str(response.choices[0].message.content)
+
+    async def request_embedding(self, str_input: str) -> list[float]:
+        response = self.client.embeddings.create(
+            model="text-embedding-ada-002", input=str_input, encoding_format="float"
+        )
+        return response.data[0].embedding
+
+    async def generate_townhall_discussion(
+        self, realm_id: int, townhall_summaries: list[str], npc_list: list[Villager], events: list[StoredEvent]
+    ) -> str:
+        realms = Realms.instance()
```

Comment:
can we initialize the class as having a realms instance, as opposed to constructing one per townhall gen req?

Reply:
hmm can assign the instance to the class but the Realms class is a singleton so it's simply getting the object, maybe creating a copy of the same instance, but it's not creating a new one

Reply:
re: singleton

having the singleton passed by init might make it easier to test for mocking

Reply:
Yeah true will change